internal/unpack/unpack.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2018 Gravitational, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unpack
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type Opts struct {
+	// SkipInvalid makes unpacking skip any invalid files rather than aborting
+	// the whole unpack.
+	SkipInvalid bool
+
+	// Filter filters out files that match the given predicate.
+	Filter func(file fs.FileInfo) bool
+}
+
+// Tgz unpacks the contents of the given gzip compressed tarball under dir.
+func Tgz(r io.Reader, dir string, opt Opts) error {
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		return err
+	}
+	return Tar(gzr, dir, opt)
+}
+
+// Tar unpacks the contents of the specified tarball under dir.
+func Tar(r io.Reader, dir string, opt Opts) error {
+	tr := tar.NewReader(r)
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+
+		if opt.Filter != nil && !opt.Filter(header.FileInfo()) {
+			continue
+		}
+
+		err = sanitizeTarPath(header, dir)
+		if err != nil {
+			if opt.SkipInvalid {
+				continue
+			}
+			return err
+		}
+
+		err = extractFile(tr, header, dir)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// extractTarFile extracts a single file or directory from tarball into dir.
+func extractFile(tr *tar.Reader, h *tar.Header, dir string) error {
+	switch h.Typeflag {
+	case tar.TypeDir:
+		return withDir(filepath.Join(dir, h.Name), nil)
+	case tar.TypeBlock, tar.TypeChar, tar.TypeReg, tar.TypeRegA, tar.TypeFifo:
+		return writeFile(filepath.Join(dir, h.Name), tr, h.Size, h.FileInfo().Mode())
+	case tar.TypeLink:
+		return writeHardLink(filepath.Join(dir, h.Name), filepath.Join(dir, h.Linkname))
+	case tar.TypeSymlink:
+		return writeSymbolicLink(filepath.Join(dir, h.Name), h.Linkname)
+	}
+	return nil
+}
+
+// sanitizeTarPath checks that the tar header paths resolve to a subdirectory
+// path, and don't contain file paths or links that could escape the tar file
+// like ../../etc/password.
+func sanitizeTarPath(h *tar.Header, dir string) error {
+	// Sanitize all tar paths resolve to within the destination directory.
+	destPath := filepath.Join(dir, h.Name)
+	if !strings.HasPrefix(destPath, filepath.Clean(dir)+string(os.PathSeparator)) {
```

Comment:
```suggestion
	cleanedDestPath := filepath.Join(dir, h.Name)
	if !strings.HasPrefix(cleanedDestPath, filepath.Clean(dir)+string(os.PathSeparator)) {
```

The fact that `filepath.Join` calls `filepath.Clean` is important for correctness here (and may not be obvious to someone reading the code), so it would help to emphasize that at least in the variable name, if not with a short comment as well.

Reply:
2fa72af5c06d8882fadef46cda1bf4f27254d793